File: csharp/DCbor/DCbor/Diag.cs
namespace BlockchainCommons.DCbor;

/// <summary>
/// Options controlling diagnostic notation formatting.
/// </summary>
public class DiagFormatOptions
{
    public bool Annotate { get; set; }
    public bool Summarize { get; set; }
    public bool Flat { get; set; }
    public TagsStoreOption Tags { get; set; } = TagsStoreOption.DefaultGlobal;

    public DiagFormatOptions WithAnnotate(bool v) { Annotate = v; return this; }
    public DiagFormatOptions WithSummarize(bool v) { Summarize = v; if (v) Flat = true; return this; }
    public DiagFormatOptions WithFlat(bool v) { Flat = v; return this; }
    public DiagFormatOptions WithTags(TagsStoreOption v) { Tags = v; return this; }
}

/// <summary>
/// Diagnostic notation rendering for CBOR values.
/// </summary>
public static class CborDiag
{
    /// <summary>Returns diagnostic notation for this CBOR value.</summary>
    public static string Diagnostic(this Cbor cbor)
    {
        return cbor.DiagnosticOpt(new DiagFormatOptions());
    }

    /// <summary>Returns diagnostic notation with annotations.</summary>
    public static string DiagnosticAnnotated(this Cbor cbor)
    {
        return cbor.DiagnosticOpt(new DiagFormatOptions { Annotate = true });
    }

    /// <summary>Returns flat (single-line) diagnostic notation.</summary>
    public static string DiagnosticFlat(this Cbor cbor)
    {
        return cbor.DiagnosticOpt(new DiagFormatOptions { Flat = true });
    }

    /// <summary>Returns a summarized representation using tag summarizers.</summary>
    public static string Summary(this Cbor cbor)
    {
        return cbor.DiagnosticOpt(new DiagFormatOptions().WithSummarize(true));
    }

    /// <summary>Returns diagnostic notation with custom options.</summary>
    public static string DiagnosticOpt(this Cbor cbor, DiagFormatOptions opts)
    {
        var item = BuildDiagItem(cbor, opts);
        return item.Format(opts);
    }

    private static DiagItem BuildDiagItem(Cbor cbor, DiagFormatOptions opts)
    {
        switch (cbor.Case)
        {
            case CborCase.UnsignedCase:
            case CborCase.NegativeCase:
            case CborCase.ByteStringCase:
            case CborCase.TextCase:
            case CborCase.SimpleCase:
                return new DiagItem.Leaf(cbor.ToString());

            case CborCase.ArrayCase a:
            {
                var items = a.Value.Select(x => BuildDiagItem(x, opts)).ToList();
                return new DiagItem.Group("[", "]", items, false, null);
            }

            case CborCase.MapCase m:
            {
                var items = new List<DiagItem>();
                foreach (var (key, value) in m.Value)
                {
                    items.Add(BuildDiagItem(key, opts));
                    items.Add(BuildDiagItem(value, opts));
                }
                return new DiagItem.Group("{", "}", items, true, null);
            }

            case CborCase.TaggedCase tg:
            {
                if (opts.Summarize)
                {
                    CborSummarizer? summarizer = null;
                    switch (opts.Tags)
                    {
                        case TagsStoreOption.Custom custom:
                            summarizer = custom.Store.GetSummarizer(tg.Tag.Value);
                            break;
                        case TagsStoreOption.Global:
                            summarizer = GlobalTags.WithTags(store => store.GetSummarizer(tg.Tag.Value));
                            break;
                    }

                    if (summarizer != null)
                    {
                        try
                        {
                            string summary = summarizer(tg.Item, opts.Flat);
                            return new DiagItem.Leaf(summary);
                        }
                        catch (Exception ex)
                        {
                            return new DiagItem.Leaf($"<error: {ex.Message}>");
                        }
                    }
                }

                string? comment = null;
                if (opts.Annotate)
                {
                    comment = opts.Tags switch
                    {
                        TagsStoreOption.Custom custom =>
                            custom.Store.AssignedNameForTag(tg.Tag),
                        TagsStoreOption.Global =>
                            GlobalTags.WithTags(store => store.AssignedNameForTag(tg.Tag)),
                        _ => null,
                    };
                }

                var child = BuildDiagItem(tg.Item, opts);
                string begin = tg.Tag.Value.ToString() + "(";
                return new DiagItem.Group(begin, ")", new List<DiagItem> { child }, false, comment);
            }

            default:
                throw new InvalidOperationException();
        }
    }
}

/// <summary>
/// Internal tree representation for diagnostic formatting.
/// </summary>
internal abstract class DiagItem
{
    internal sealed class Leaf : DiagItem
    {
        internal string Text { get; }
        internal Leaf(string text) { Text = text; }
    }

    internal sealed class Group : DiagItem
    {
        internal string Begin { get; }
        internal string End { get; }
        internal List<DiagItem> Items { get; }
        internal bool IsPairs { get; }
        internal string? Comment { get; }

        internal Group(string begin, string end, List<DiagItem> items, bool isPairs, string? comment)
        {
            Begin = begin;
            End = end;
            Items = items;
            IsPairs = isPairs;
            Comment = comment;
        }
    }

    internal string Format(DiagFormatOptions opts) => FormatOpt(0, "", opts);

    private string FormatOpt(int level, string separator, DiagFormatOptions opts)
    {
        switch (this)
        {
            case Leaf leaf:
                return FormatLine(level, opts, leaf.Text, separator, null);

            case Group g:
                if (!opts.Flat
                    && (ContainsGroup()
                        || TotalStringsLen() > 20
                        || GreatestStringsLen() > 20))
                {
                    return MultilineComposition(level, separator, opts);
                }
                return SingleLineComposition(level, separator, opts);

            default:
                throw new InvalidOperationException();
        }
    }

    private static string FormatLine(int level, DiagFormatOptions opts, string text, string separator, string? comment)
    {
        string indent = opts.Flat ? "" : new string(' ', level * 4);
        string result = $"{indent}{text}{separator}";
        if (comment != null)
            return $"{result}   / {comment} /";
        return result;
    }

    private string SingleLineComposition(int level, string separator, DiagFormatOptions opts)
    {
        switch (this)
        {
            case Leaf leaf:
                return FormatLine(level, opts, leaf.Text, separator, null);

            case Group g:
            {
                var components = g.Items.Select(item => item switch
                {
                    Leaf l => l.Text,
                    Group => item.SingleLineComposition(level + 1, separator, opts),
                    _ => throw new InvalidOperationException(),
                }).ToList();

                string pairSeparator = g.IsPairs ? ": " : ", ";
                string joined = JoinItems(components, ", ", pairSeparator);
                string text = g.Begin + joined + g.End;
                string? comment = g.Comment;
                return FormatLine(level, opts, text, separator, comment);
            }
            default:
                throw new InvalidOperationException();
        }
    }

    private string MultilineComposition(int level, string separator, DiagFormatOptions opts)
    {
        switch (this)
        {
            case Leaf leaf:
                return leaf.Text;

            case Group g:
            {
                var lines = new List<string>();
                // Use a non-flat version for the opening line
                var nonFlatOpts = new DiagFormatOptions
                {
                    Annotate = opts.Annotate,
                    Summarize = opts.Summarize,
                    Flat = false,
                    Tags = opts.Tags,
                };
                lines.Add(FormatLine(level, nonFlatOpts, g.Begin, "", g.Comment));

                for (int i = 0; i < g.Items.Count; i++)
                {
                    string sep;
                    if (i == g.Items.Count - 1)
                        sep = "";
                    else if (g.IsPairs && (i & 1) == 0)
                        sep = ":";
                    else
                        sep = ",";
                    lines.Add(g.Items[i].FormatOpt(level + 1, sep, opts));
                }

                lines.Add(FormatLine(level, opts, g.End, separator, null));
                return string.Join("\n", lines);
            }
            default:
                throw new InvalidOperationException();
        }
    }

    private int TotalStringsLen()
    {
        return this switch
        {
            Leaf leaf => leaf.Text.Length,
            Group g => g.Items.Sum(item => item.TotalStringsLen()),
            _ => 0,
        };
    }

    private int GreatestStringsLen()
    {
        return this switch
        {
            Leaf leaf => leaf.Text.Length,
            Group g => g.Items.Count > 0 ? g.Items.Max(item => item.TotalStringsLen()) : 0,
            _ => 0,
        };
    }

    private bool IsGroup() => this is Group;

    private bool ContainsGroup()
    {
        return this switch
        {
            Leaf => false,
            Group g => g.Items.Any(x => x.IsGroup()),
            _ => false,
        };
    }

    private static string JoinItems(List<string> elements, string itemSeparator, string? pairSeparator)
    {
        pairSeparator ??= itemSeparator;
        var sb = new System.Text.StringBuilder();
        for (int i = 0; i < elements.Count; i++)
        {
            sb.Append(elements[i]);
            if (i != elements.Count - 1)
            {
                sb.Append((i & 1) != 0 ? itemSeparator : pairSeparator);
            }
        }
        return sb.ToString();
    }
}

File: python/dcbor/src/dcbor/diag.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .string_util import flanked

if TYPE_CHECKING:
    from .cbor import CBOR
    from .tags_store import CBORSummarizer, TagsStore

_NO_TAGS = object()


class _DiagItem:
    pass


class _DiagLeaf(_DiagItem):
    __slots__ = ("text",)

    def __init__(self, text: str) -> None:
        self.text = text


class _DiagGroup(_DiagItem):
    __slots__ = ("begin", "end", "items", "is_pairs", "comment")

    def __init__(
        self,
        begin: str,
        end: str,
        items: list[_DiagItem],
        is_pairs: bool,
        comment: str | None,
    ) -> None:
        self.begin = begin
        self.end = end
        self.items = items
        self.is_pairs = is_pairs
        self.comment = comment


def _total_strings_len(item: _DiagItem) -> int:
    if isinstance(item, _DiagLeaf):
        return len(item.text)
    if isinstance(item, _DiagGroup):
        return sum(_total_strings_len(child) for child in item.items)
    return 0


def _greatest_strings_len(item: _DiagItem) -> int:
    if isinstance(item, _DiagLeaf):
        return len(item.text)
    if isinstance(item, _DiagGroup):
        if not item.items:
            return 0
        return max(_total_strings_len(child) for child in item.items)
    return 0


def _is_group(item: _DiagItem) -> bool:
    return isinstance(item, _DiagGroup)


def _contains_group(item: _DiagItem) -> bool:
    if isinstance(item, _DiagGroup):
        return any(_is_group(child) for child in item.items)
    return False


def _joined(elements: list[str], item_separator: str, pair_separator: str | None) -> str:
    pair_sep = pair_separator if pair_separator is not None else item_separator
    result: list[str] = []
    length = len(elements)
    for i, elem in enumerate(elements):
        result.append(elem)
        if i != length - 1:
            if i & 1 != 0:
                result.append(item_separator)
            else:
                result.append(pair_sep)
    return "".join(result)


def _format_line(
    level: int,
    flat: bool,
    string: str,
    separator: str,
    comment: str | None,
) -> str:
    indent = "" if flat else " " * (level * 4)
    result = f"{indent}{string}{separator}"
    if comment is not None:
        return f"{result}   / {comment} /"
    return result


def _single_line(
    item: _DiagItem,
    level: int,
    separator: str,
    flat: bool,
) -> str:
    if isinstance(item, _DiagLeaf):
        return _format_line(level, flat, item.text, separator, None)
    if isinstance(item, _DiagGroup):
        components: list[str] = []
        for child in item.items:
            if isinstance(child, _DiagLeaf):
                components.append(child.text)
            else:
                components.append(_single_line(child, level + 1, separator, flat))
        pair_sep = ": " if item.is_pairs else ", "
        body = _joined(components, ", ", pair_sep)
        text = flanked(body, item.begin, item.end)
        return _format_line(level, flat, text, separator, item.comment)
    return ""


def _multiline(
    item: _DiagItem,
    level: int,
    separator: str,
    flat: bool,
) -> str:
    if isinstance(item, _DiagLeaf):
        return item.text
    if isinstance(item, _DiagGroup):
        lines: list[str] = []
        lines.append(
            _format_line(level, False, item.begin, "", item.comment)
        )
        for i, child in enumerate(item.items):
            if i == len(item.items) - 1:
                sep = ""
            elif item.is_pairs and i & 1 == 0:
                sep = ":"
            else:
                sep = ","
            lines.append(_format_item(child, level + 1, sep, flat))
        lines.append(_format_line(level, flat, item.end, separator, None))
        return "\n".join(lines)
    return ""


def _format_item(
    item: _DiagItem,
    level: int,
    separator: str,
    flat: bool,
) -> str:
    if isinstance(item, _DiagLeaf):
        return _format_line(level, flat, item.text, separator, None)
    if not flat and (
        _contains_group(item)
        or _total_strings_len(item) > 20
        or _greatest_strings_len(item) > 20
    ):
        return _multiline(item, level, separator, flat)
    return _single_line(item, level, separator, flat)


def _build_diag_item(
    cbor: CBOR,
    annotate: bool,
    summarize: bool,
    tags_store: TagsStore | None,
) -> _DiagItem:
    from .cbor import CBORCase

    match cbor.case:
        case (
            CBORCase.UNSIGNED
            | CBORCase.NEGATIVE
            | CBORCase.BYTE_STRING
            | CBORCase.TEXT
            | CBORCase.SIMPLE
        ):
            return _DiagLeaf(cbor._leaf_str())

        case CBORCase.ARRAY:
            items = [
                _build_diag_item(child, annotate, summarize, tags_store)
                for child in cbor.value
            ]
            return _DiagGroup("[", "]", items, False, None)

        case CBORCase.MAP:
            items: list[_DiagItem] = []
            for key, value in cbor.value.iter():
                items.append(_build_diag_item(key, annotate, summarize, tags_store))
                items.append(_build_diag_item(value, annotate, summarize, tags_store))
            return _DiagGroup("{", "}", items, True, None)

        case CBORCase.TAGGED:
            tag, item = cbor.value

            if summarize:
                summarizer = _get_summarizer(tag.value, tags_store)
                if summarizer is not None:
                    try:
                        summary_text = summarizer(item, True)
                        return _DiagLeaf(summary_text)
                    except Exception as e:
                        return _DiagLeaf(f"<error: {e}>")

            comment: str | None = None
            if annotate:
                comment = _get_assigned_name(tag, tags_store)

            child_item = _build_diag_item(item, annotate, summarize, tags_store)
            begin = str(tag.value) + "("
            return _DiagGroup(begin, ")", [child_item], False, comment)

    return _DiagLeaf("")


def _get_summarizer(tag_value: int, tags_store: TagsStore | None) -> CBORSummarizer | None:
    if tags_store is None:
        from .tags_store import _get_global_tags
        store = _get_global_tags()
        return store.summarizer(tag_value)
    if tags_store is _NO_TAGS:
        return None
    return tags_store.summarizer(tag_value)


def _get_assigned_name(tag: object, tags_store: TagsStore | None) -> str | None:
    if tags_store is None:
        from .tags_store import _get_global_tags
        store = _get_global_tags()
        return store.assigned_name_for_tag(tag)
    if tags_store is _NO_TAGS:
        return None
    return tags_store.assigned_name_for_tag(tag)


def diagnostic_impl(
    cbor: CBOR,
    *,
    annotate: bool = False,
    flat: bool = False,
    summarize: bool = False,
    tags_store: TagsStore | None = None,
) -> str:
    if summarize:
        flat = True
    item = _build_diag_item(cbor, annotate, summarize, tags_store)
    return _format_item(item, 0, "", flat)
